File: valid BST_iterative .py

# ITERATIVE SOLUTION / DFS APPROACH

#  ===> Solution : BETTER than creating a list out of  the inorder traversal of tree and then checking if it is in increasing order as this solution returns False immediately when a breach happens.

# Time Complexity: O(N), WORST CASE of perfect BST, it has to traverse through all nodes to return True
# Space Complexity: No additional space, except the Recursive stack
# Did problem run on Leetcode: yes

class TreeNode:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None

class Solution:

    def isValidBST(self, root):
        return self.helper(root, None, None)

    def helper(self, root, min, max):
        # BASE CASE
        if not root:
            return True

        # LOGIC
        if min != None and root.key <= min:
            return False
        if max != None and root.key >= max:
            return False
        return self.helper(root.left, min, root.key) and self.helper(root.right, root.key, max)
        


a = TreeNode(20)
a.left = TreeNode(15)
a.right = TreeNode(25)
a.left.left = TreeNode(13)
a.left.right = TreeNode(18)
a.right.left = TreeNode(24)
a.right.right = TreeNode(27)
a.left.left.left = TreeNode(10)
a.left.left.right = TreeNode(14)
a.left.right.left = TreeNode(16)
a.left.right.right = TreeNode(19)

obj = Solution()
print(obj.isValidBST(root = a))




            
        

File: Valid BST Iterative.java
// Time Complexity : O(n)
// Space Complexity : O(h)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    public boolean isValidBST(TreeNode root) {
        
        // Intialize stack
        Stack<TreeNode> st = new Stack<>();

        // Intialize previous pointer to compare value with
        TreeNode prev = null;
        
        // Iterate through the tree
        while(root != null || !st.isEmpty()){        
            
            // keep pushing the left children nodes
            // keep updating the root until we hit null value
            while(root != null){
                st.push(root);
                root = root.left;
            }

            // once we hit the null value on the left side,
            // pop the element from the stack

            root = st.pop();
        
            // check if the previously popped root is smaller than the current root
            if(prev != null && prev.val >= root.val){
                return false;
            }
            
            // update previous pointer
            prev  = root;

            // do the same for children nodes on right
            root = root.right;  
        }
        
        return true;
    }
}